ReleaseProcessScript/tools/Core/git_base_functions.ps1

```diff
@@ -22,14 +22,29 @@ function Get-Current-Branchname ()
   return $(git symbolic-ref --short -q HEAD)
 }
 
-function Get-Last-Version-Of-Branch-From-Tag-Exists ($Branchname)
+function Get-Last-Version-Of-Branch-From-Tag-Exists ($From, $To)
 {
-  return (git describe $Branchname --match "v[0-9]*" --abbrev=0 2>&1) -and $?
+  $params = Get-Parameters-To-Retrieve-Last-Version-From-Tag $From $To
+  return (git $params[0] for-each-ref $params[1] 2>&1) -and $?
 }
 
-function Get-Last-Version-Of-Branch-From-Tag ($Branchname)
+function Get-Last-Version-Of-Branch-From-Tag ($From, $To)
 {
-  return git describe $Branchname --match "v[0-9]*" --abbrev=0
+  $params = Get-Parameters-To-Retrieve-Last-Version-From-Tag $From $To
+  return git $params[0] for-each-ref $params[1]
+}
+
+function Get-Parameters-To-Retrieve-Last-Version-From-Tag ($From, $To) 
+{
+  #pre-release versions are sorted in order of specification
+  $gitParams = @("-c", "versionsort.suffix=-alpha", "-c", "versionsort.suffix=-beta", "-c", "versionsort.suffix=-rc")
+  $refParams = @("refs/tags", "--sort=-version:refname", "--format=`"%(refname:short)`"", "--count=1")
+  
+  #version tags are looked for starting from revision $From (backwards in history) until $To (empty string is equivalent to "unbounded")
+  $refParams += If ([string]::IsNullOrEmpty($From)) {"--merged=HEAD"} Else {"--merged=`"$From`""}
+  $refParams += If ([string]::IsNullOrEmpty($To)) {""} Else {"--contains=`"$To`""}
+
+  return @($gitParams, $refParams)
```

Comment:
This is a noteworthy change I would like to spend some words on. The previous approach proved unreliable in certain scenarios because tags that were not merged into the current branch (e. g. from a support branch) were considered "closer" than tags from a merged branch, despite them being more recent (e. g. from a prerelease branch).

The example below demonstrates the issue. It might be a bit contrived, but hopefully shows the essence of the problem.

Assume the following repository (I omitted the hotfix/v2.28.1 branch for the sake of brevity):
```
PS> git log --all --decorate --oneline --graph
*   fdd2b09 (HEAD -> hotfix/v2.28.2) Merge branch 'prelease/v2.28.2-alpha.1' into hotfix/v2.28.2
|\
| * 5b8e87b (tag: v2.28.2-alpha.1, prelease/v2.28.2-alpha.1) Update metadata to version '2.28.2-alpha.1'.
|/
* 97935fa Feature-C
* 7dd3a53 (tag: v2.28.1, support/v2.28) Feature-B
* 31777be Feature-A
* a7360b4 (tag: v2.28.0, master) Initial Release
```

The previous approach returns the tag from commit "Feature-B". However, `v2.28.2-alpha.1` would be correct.
```
PS> git describe --match "v[0-9]*" --abbrev=0
v2.28.1
```

The proposed solution relies less on the DAG structure of the git log (as it did until now), but more on version numbers that are strictly monotonically increasing. The `--sort=-version:refname` sorts all version tags such that the newest version is the first line in the output. This assumption is hopefully a reasonable compromise.

```
PS> git -c "versionsort.suffix=-alpha" -c "versionsort.suffix=-beta" -c "versionsort.suffix=-rc" -c "versionsort.suffix=`"`"" for-each-ref refs/tags --sort=-version:refname --format="%(refname:short)" --count=1
v2.28.2-alpha.1
```

Reply:
Hmm... I just tried to understand this and google / stackoverflow suggests:
`git describe`

And it does seem to work and print out the correct tags. tested it in hotfix/2.28 and on develop, but we should do more testing.

https://stackoverflow.com/questions/33275853/git-find-all-tags-reachable-from-a-commit
https://git-scm.com/docs/git-describe
> The command finds the most recent tag that is reachable from a commit

Could you please check if this is a simpler/better solution which also gives us the correct results? Disclaimer: you may have to do a string-split on the result because it prints
> v2.28.0-alpha.3-3-g3a9b7c46

But there could very well be a parameter to make this all better out of the box.

Reply:
Right at the start: Sorry for the lengthy answer.

Yes, I was able to to retrieve the correct tag in re-motion as well - after all, the function has been there in this form for quite some time. It's just that, when testing with different combinations of branches and tags (e. g. do not fall back to a tag from the support branch or even master), I noticed potentially unwanted behaviour.

For example, take the log depicted in the screenshot below (see [git commands](https://github.com/re-motion/SharedSource-Build/blob/1ef9a8d05a6eda22c12a8c84c878fbcc7e723758/ReleaseProcessScript/tools/PesterTester/GitCommandsForReferenceDir/Hotfix-PreRelease-Beta/Initialize.ps1) from the respective tests; "First commit", "Second commit" including 1.0.0 and 1.1.0 tags are created by the infrastructure beforehand).
![image](https://user-images.githubusercontent.com/40166342/107922171-ec36cf00-6f6f-11eb-8785-761dc27e1045.png)

The `Get-Last-Version-Of-Branch-From-Tag` should return `v2.28.1-beta.1` in this case, exemplified by the corresponding test.
https://github.com/re-motion/SharedSource-Build/blob/1ef9a8d05a6eda22c12a8c84c878fbcc7e723758/ReleaseProcessScript/tools/PesterTester/IntegrationTests.Tests.ps1#L80-L86

However, I was often (not always, that's the great thing I don't understand) greeted with a test failure because the function returned the `v1.1.0` tag.

```
PS> git describe --match "v[0-9]*" --abbrev=0
v1.1.0

PS> git describe --match "v[0-9]*" --abbrev=0 --debug
describe HEAD
No exact match on refs or tags, searching to describe
finished search at c6442cf2e010806e33d20243e83b1ef5f51e7c6f
 annotated          4 v1.1.0
 annotated          5 v2.28.1-beta.1
 annotated          5 v1.0.0
traversed 6 commits
v1.1.0
```

![image](https://user-images.githubusercontent.com/40166342/107923665-18ebe600-6f72-11eb-980c-10b604d17563.png)

From what I was able to gather, the numeric values printed out with the `--debug` flag (and if you leave out ``--abrev=0``) is the "distance", i. e. the number of commits from HEAD that were traversed - in some kind of depth-first search way. Starting from the latest commit, it's four hops to the v1.1.0 tag (then backtracking) and another one to the v2.28.1-beta.1 tag. However, that was just an assumption.

The "nice" thing is that I was not able to find information that fully describe the algorithm employed by git. The behaviour above is not consistent to a degree of 100 %. There is a failed Shared Source Build build of this 151-branch where the test was successful - however, an analogous test with the a very similar repo structure failed. It was an alpha.2 release with a previous alpha.1 release and the function returned alpha.1 as latest version instead of alpha.2, which is even more strange in my opinion - see screenshot below for the repository.

![image](https://user-images.githubusercontent.com/40166342/107925839-425a4100-6f75-11eb-98ab-2b4233f38384.png)

In summary, I just don't fully understand, which made me uncomfortable not to change anything. Since I was not able to find explanations for this inconsistent behaviour (different outputs for the very same repository on different executions), I took it as the function being not as reliable as we might want it to be as it relies on measuring "closeness" in a DAG, which can be tricky and may have different notions depending on what to achieve. That's why the alternative I proposed solely relies on the order relation that's defined for semver versions - in other words we have a more lengthy command, but more predictable results.

If such "problematic" cases are not relevant in practice, I can gladly retract the changes and stick to the old solution. I am not even sure if this undesirable behaviour only occurs in repositories as small as in the test cases I created locally - "real" repositories have quite different distances between version tags.

Apologies again for my extensive answer. Hopefully, I was able to explain my line of thinking in a digestible, linear way. Nevertheless, further discussion might be better lead in person.

Reply:
@raffaelfoidl 
Did you do annotated tags or just tags?

And yes, abbrev is weird. Here's the docs
> Instead of using the default 7 hexadecimal digits as the abbreviated object name, use <n> digits, or as many digits as needed to form a unique object name. **An <n> of 0 will suppress long format, only showing the closest tag.**

So, `abbrev=0` seems to be what we need. 

Yes, a live session would be helpful.

Reply:
I used annotated tags.

Reply:
annotated: okay, that's not it, then.
Another question:
> The command finds the most recent tag that is reachable from a commit. 

The docs refer to the "most recent tag".  Is it possible that you created the tags in the wrong temporal order?

Reply:
Unfortunately, the temporal order should be correct (from the command listings and the log):
```
v1.0.0:                10:04:21
v1.1.0:                10:04:21
v2.28.1-alpha.1: 10:04:23
v2.28.1-alpha.2: 10:04:23
```

Funny enough, if I add `--candidates=1` to the command, it seems to return the correct result.
> Instead of considering only the 10 most recent tags as candidates to describe the input commit-ish consider up to <n> candidates. Increasing <n> above 10 will take slightly longer but may produce a more accurate result. An <n> of 0 will cause only exact matches to be output.

```
PS> git describe --abbrev=0 --debug
describe HEAD
No exact match on refs or tags, searching to describe
finished search at ae05250e1557b88455b9db08289ce15ce09ac1a9
 annotated          7 v1.1.0
 annotated          8 v2.28.1-alpha.2
 annotated          8 v2.28.1-alpha.1
 annotated          8 v1.0.0
traversed 9 commits
v1.1.0

PS> git describe --candidates=1 --abbrev=0 --debug
describe HEAD
No exact match on refs or tags, searching to describe
 annotated          8 v2.28.1-alpha.2
traversed 9 commits
more than 1 tags found; listed 1 most recent
gave up search at 07896fb7b383d61a015f8390b11eaf3634b1ec25
v2.28.1-alpha.2
```

With this addition, the tests are green, even for many executions. I am a bit surprised that the revision after `gave up search` is not the one with the output-tag, but what would have been the second recent tag (`v2.28.1-alpha.1`).

So even though that seems to work, I am anxious as I do not fully understand what is going on. Anyway, with `git describe` I would also still have to figure out how stop the search at the point from which the hotfix branch started (i. e. tags from support/master should not even be an option, for the case that there is no tag on the hotfix branch yet).

Reply:
About stopping: we could just use pattern logic here and match all tags that fit the branch name's version. that way, we could just look at all the tags and would still get the correct result.

One other option about the "recent" maybe there's a timestamp issue, i.e. tags created within the same second are ambiguous.